### PR TITLE
Coverage

### PR DESCRIPTION
**Pull request by Thrameos**
This is the last coverage PR before the next release.  This PR should pin down the last of the exposed API so that we can detect API breakage going forward.

Issues addressed in this PR:
- Removed stub method attachToJVM.  This never served a useful purpose.   When we start Python from within Java it will be automatically attached so this never served a function.
- Exposed registerJVMInitializer.  This internal method have been used to more or less extent in JPype for a while.  Currently it has been replaced with calls in the core but it is still a necessary tool for external libraries that need to create resources when the JVM starts.  I would recommend renaming it to ``onJVMStart`` to be more consistent with other exported symbols (and less jargony)
- Switched dependency tracker to ivy.   We are going to need better dependencies in the next version so this is preemptive. 

Stuff that remains broken, deprecated, or bad ideas:
- JString was a factory from long ago introduced when we did not have easy imports.   It currently is a psuedo alias for java.lang.String but lacks any methods.  
- JException is a psuedo alias for java.lang.Throwable.  It is badly named and has little functionality.
- JObject(x, PythonType) this is a legacy call which will interfere with functionality going forward and is marked deprecated.
- [x] **FIXED** JObject() without an argument.  For reasons that are a complete mystery (more like implementation bug), it appears this creates a null pointer rather than a java.lang.Object instance.  I would recommend correcting this one as it is very counter intuitive and basically no different from JObject(None).
- beans.py This is a legacy function which remains as dangerous as ever.
- [x] **FIXED** Memory compiler - this is an undocumented feature which was part of the earliest attempts at extending Java classes.  Unfortunately, only some JDK configurations support it.  I recommend pulling it as not useful.
- The bytebuffer api should have been moved to a customize wrapper for ByteBuffer just like threads so that we don't need a separate API call for it.

What is uncovered by tests:
- Glob patterns in classpath.   This one as written is not testable, but the code pattern is hit often enough that it is unlikely to remain broken.  (in getClassPath and core)
- Sections of customizers.  Some patterns are challenging to hit as they can be hit before and after JVM start.  But there no chance of API slipage in this section.
- jvmFinder - this is untestable without a whole lot of mock.  It is hard to test the api independent of the implementation, but if this is seriously broken it should show in the totality of the test bench.
- beans.py if it is broken you get to keep both pieces.
- import hooks was meant to allow a user to write an import hook that pulls python code from a jar.  To my knowledge no one has accomplished this task, so we don't have tests for it.  We may want to just shelve that section as it is likely not correct in the current form.
- pickle module - there is a single section of code which is for Python 2 and Python 3 running with the pure Python implementation.  It was known good when we ran Python 2 tests but is currently not checked.  As it still possible to exercise in 3 it has been left as is.

CPython sections not covered:
- [x] pyjp_array - error paths that are difficult to reach
- [ ] pyjp_array - buffer configurations that require specialized memory layouts
- [ ] pyjp_buffer - error handling, special memory requests and non-writable buffers.
- [x] pyjp_char - wide character paths, error handling, certain comparison cases.
- [ ] pyjp_class - traverse and cleanup routines (the module does not shutdown in such a way that these get called).  
- [ ] pyjp_class - get/set with non unicode.  These paths are sanity checks that Python has catches for before it gets there.  It is possible to hit by explicitly calling from with Python so the checks should remain.
- [x] pyjp_class canCast, new private API for dbapi2.  **should be added to coverage**
- [x] pyjp_classhints - exclusions on conversions **should be added to  coverage**
- [ ] pyjp_field - clean up routines (if the classes don't go away the fields cant as well)
- [x] pyjp_method matches = new routine **should be added to coverage**
- [ ] pyjp_module - error handling routines
- [ ] pyjp_module - buffer routines for odd memory layouts
- [ ] pyjp_number - error handling routines
- [x] pyjp_object - error handling routines
- [x] pyjp_package - error handling routines
- [x] pyjp_package - error redirection (Not sure how this is missed as it is used in imports)
- [x] pyjp_value - error handling routines

@marscher If you can be so kind as to look over this big list and flag anything you would like dealt with for the next release (or if you have some instructions on how it should be resolved).  I will get them taken care of over the next few days.

### CODE REVIEW

File: jpype/_jcustomizer.py [hunk 1]
```diff
@@ -233,6 +233,12 @@ def applyCustomizers(self, name, bases, members):
         for b in self.bases:
             bases.insert(0, b)
 
+        try:
+            module = name.rsplit('.', 1)[0]
+            members['__module__'] = module
+        except Exception:
```
[the remark this answers is not in the document]
Comment (Thrameos) [reply jpype/_jcustomizer.py:239]: Likely.   I need it to swallow the cases 

- no dot in the name
- something wierd like "$" in the name

I can run through and make sure that I limit it to one of those cases.

[the remark this answers is not in the document]
Comment (Thrameos) [reply jpype/_jcustomizer.py]: Okay it appears that I don't need the try here.  I just need to check the length.

**Comment by Thrameos** [comment]
Assuming that everything is ready to start a release candidate, I am going to branch for the next  development cycle as I should be able to go a while before the core diverges such that I can't merge master in.  Estimated time to next prototype version will be about 3 months.

File: test/jpypetest/test_hints.py [hunk 2]
```diff
@@ -389,3 +389,21 @@ def testInstant(self):
         self.assertTrue(datetime.datetime in hints.implicit)
         self.assertTrue(len(hints.explicit) == 0)
         self.assertTrue(len(hints.none) == 0)
+
+    def testAddTypeBad(self):
```
Comment (marscher) [inline test/jpypetest/test_hints.py:393]: is this testing of private api as all the methods/props involved begin with an underscore?

Comment (Thrameos) [reply test/jpypetest/test_hints.py:393]: Well sort of.  I could have tried to hit the error check branches using the ``@JConversion`` dispatch function, but not every path is easily hit through the dispatch.  I am not sure it is worth trying to find the code that will fool the dispatch into applying the customizer to a non-function.

**Pull request merged**